contracts/ConvictionVoting.sol

```diff
@@ -305,124 +380,155 @@ contract ConvictionVoting is AragonApp, TokenManagerHook {
-            proposal.convictionLast,
+            blockNumber - _proposal.blockLast, // we assert it doesn't overflow above
+            _proposal.convictionLast,
             _oldStaked
         );
-        proposal.blockLast = blockNumber;
-        proposal.convictionLast = conviction;
+        _proposal.blockLast = blockNumber;
+        _proposal.convictionLast = conviction;
     }
 
     /**
      * @dev Stake an amount of tokens on a proposal
-     * @param _id Proposal id
+     * @param _proposalId Proposal id
      * @param _amount Amount of staked tokens
      * @param _from Account from which we stake
      */
-    function _stake(uint256 _id, uint256 _amount, address _from) internal {
+    function _stake(uint256 _proposalId, uint256 _amount, address _from) internal {
+        Proposal storage proposal = proposals[_proposalId];
         require(_amount > 0, ERROR_AMOUNT_CAN_NOT_BE_ZERO);
-        uint256 unstaked = stakeToken.balanceOf(_from).sub(stakesPerVoter[_from]);
-        if (_amount > unstaked) {
-            _withdrawUnstakedTokens(_amount.sub(unstaked), _from, true);
+        require(proposal.proposalStatus == ProposalStatus.Active, ERROR_PROPOSAL_NOT_ACTIVE);
+
+        uint256 unstakedAmount = stakeToken.balanceOf(_from).sub(totalVoterStake[_from]);
+        if (_amount > unstakedAmount) {
+            _withdrawInactiveStakedTokens(_amount.sub(unstakedAmount), _from);
         }
-        // make sure user does not stake more than she has
-        require(stakesPerVoter[_from].add(_amount) <= stakeToken.balanceOf(_from), ERROR_STAKED_MORE_THAN_OWNED);
-        Proposal storage proposal = proposals[_id];
-        uint256 oldStaked = proposal.stakedTokens;
+
+        require(totalVoterStake[_from].add(_amount) <= stakeToken.balanceOf(_from), ERROR_STAKING_MORE_THAN_AVAILABLE);
+
+        uint256 previousStake = proposal.stakedTokens;
         proposal.stakedTokens = proposal.stakedTokens.add(_amount);
-        proposal.stakesPerVoter[_from] = proposal.stakesPerVoter[_from].add(_amount);
-        stakesPerVoter[_from] = stakesPerVoter[_from].add(_amount);
+        proposal.voterStake[_from] = proposal.voterStake[_from].add(_amount);
+        totalVoterStake[_from] = totalVoterStake[_from].add(_amount);
+        totalStaked = totalStaked.add(_amount);
+
         if (proposal.blockLast == 0) {
             proposal.blockLast = getBlockNumber64();
+        } else {
+            _calculateAndSetConviction(proposal, previousStake);
+        }
+
+        _updateVoterStakedProposals(_proposalId, msg.sender);
+
+        emit StakeAdded(_from, _proposalId, _amount, proposal.voterStake[_from], proposal.stakedTokens, proposal.convictionLast);
+    }
+
+    function _updateVoterStakedProposals(uint256 _proposalId, address _submitter) internal {
+        uint256[] storage voterStakedProposalsArray = voterStakedProposals[_submitter];
+        require(voterStakedProposalsArray.length < MAX_STAKED_PROPOSALS, ERROR_MAX_PROPOSALS_REACHED);
```

Comment:
Ah should we move this `require` inside the condition? Otherwise i think if a user is staking to `MAX_STAKED_PROPOSALS ` proposals, it won't be able to increase his current stake on any of them.

Reply:
Excellent spot my man!